src/ipld/frozen_cids.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::utils::cid::{CidVariant, BLAKE2B256_SIZE};
+use cid::{
+    multihash::{self, Code::Blake2b256},
+    Cid,
+};
+use fvm_ipld_encoding::DAG_CBOR;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+// Similar to the `CidHashMap` implementation, `FrozenCids` optimizes storage of
+// CIDs that would normally be stored as a vector of CIDs. The V1 DAG-CBOR
+// Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid`
+// type) is +99.99% of all CIDs, so very few CIDs need to be stored in the
+// `Heap(Box<Cid>)` variant of `SmallCid`. The Box type has been chosen to make
+// `FrozenCids` explicitly immutable due to the fact that the vectors are parsed
+// from CBOR and are immutable.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FrozenCids(Box<[SmallCid]>);
+
+impl Default for FrozenCids {
+    fn default() -> Self {
+        FrozenCids(Box::new([]))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+enum SmallCid {
+    Heap(Box<Cid>),
+    Inline([u8; 32]),
+}
+
+pub struct FrozenCidsIterator<'a> {
```

Comment:
I think this should wrap https://doc.rust-lang.org/stable/std/slice/struct.Iter.html, and inflate the inline CID on iteration using `From`